core/src/main/java/com/datastrato/gravitino/storage/relational/RelationalGarbageCollector.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2024 Datastrato Pvt Ltd.
+ * This software is licensed under the Apache License version 2.
+ */
+
+package com.datastrato.gravitino.storage.relational;
+
+import static com.datastrato.gravitino.Configs.STORE_DELETE_AFTER_TIME;
+import static com.datastrato.gravitino.Configs.VERSION_RETENTION_COUNT;
+
+import com.datastrato.gravitino.Config;
+import com.google.common.annotations.VisibleForTesting;
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class RelationalGarbageCollector implements Closeable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(RelationalGarbageCollector.class);
+  private final RelationalBackend backend;
+
+  private final Config config;
+
+  @VisibleForTesting
+  final ScheduledExecutorService garbageCollectorPool =
+      new ScheduledThreadPoolExecutor(
+          2,
+          r -> {
+            Thread t = new Thread(r, "RelationalBackend-Garbage-Collector");
+            t.setDaemon(true);
+            return t;
+          },
+          new ThreadPoolExecutor.AbortPolicy());
+
+  public RelationalGarbageCollector(RelationalBackend backend, Config config) {
+    this.backend = backend;
+    this.config = config;
+  }
+
+  public void start() {
+    long dateTimeLineMinute = config.get(STORE_DELETE_AFTER_TIME) / 1000 / 60;
+
+    // We will collect garbage every 10 minutes at least. If the dateTimeLineMinute is larger than
+    // 100 minutes, we would collect garbage every dateTimeLineMinute/10 minutes.
+    long frequency = Math.max(dateTimeLineMinute / 10, 10);
+    garbageCollectorPool.scheduleAtFixedRate(this::collectAndClean, 5, frequency, TimeUnit.MINUTES);
+  }
+
+  private void collectAndClean() {
+    long threadId = Thread.currentThread().getId();
+    LOG.info("Thread {} start to collect garbage...", threadId);
+
+    try {
+      LOG.info("Start to collect and delete legacy data by thread {}", threadId);
+      long legacyTimeLine = System.currentTimeMillis() - config.get(STORE_DELETE_AFTER_TIME);
+      backend.hardDeleteLegacyData(legacyTimeLine);
```

Comment:
I think we can add a loop here to delete all expired data in a single round as soon as possible. The deletion result is obtained each time in the loop. When the deletion result is 0, the loop ends.
```
    Arrays.asList(Entity.EntityType.values()).forEach(type -> {
      long deleteCount = Long.MAX_VALUE;
      while (deleleCount > 0) {
        deleteCount = backend.hardDeleteLegacyData(type, legacyTimeLine);
      }
    });
```